### ISSUE TITLE
[no-unnecessary-condition] unexpected reporting

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#3322]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": 2
  }
}
```

```TS
import { MLRule, Ruleset } from '@markuplint/ml-core'
import { RuleConfigValue } from '@markuplint/ml-config'

export function tryRequirePkg<T>(pkg: string): T | undefined
export function tryRequirePkg<T>(pkg: string, throwError: true): T
export function tryRequirePkg<T>(
  pkg: string,
  throwError?: boolean,
): T | undefined {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
    const result = require(pkg) as T | { default: T }
    return 'default' in result ? result.default : result
  } catch (err) {
    if (throwError) {
      throw err
    }
  }
}

export function moduleAutoLoader<T extends RuleConfigValue, O = unknown>(
  ruleset: Ruleset,
) {
  const rules: Array<MLRule<T, O>> = []
  const errors: unknown[] = []

  for (const ruleName of Object.keys(ruleset.rules)) {
    let rule: MLRule<T, O> | null = null

    try {
      rule = tryRequirePkg<MLRule<T, O>>(`@markuplint/rule-${ruleName}`, true)
    } catch (e) {
      errors.push(e)
    }

    if (!rule) {
      continue
    }

    try {
      rule = tryRequirePkg<MLRule<T, O>>(`markuplint-rule-${ruleName}`, true)
    } catch (e) {
      errors.push(e)
    }

    if (!rule) { // reported here
      continue
    }

    rules.push(rule)
  }

  return {
    rules,
    errors,
  }
}

```

```jsonc
{
  "compilerOptions": {
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "importHelpers": true,
    "incremental": true,
    "jsx": "preserve",
    "lib": ["ESNext"],
    "module": "ESNext",
    "moduleResolution": "node",
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "stripInternal": true,
    "target": "ES2015"
  }
}
```

**Expected Result**

No reporting

**Actual Result**

![image](https://user-images.githubusercontent.com/8336744/116407789-b57d0f00-a864-11eb-8239-e121e979e21e.png)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.0` |
| `@typescript-eslint/parser`        | `4.22.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.25.0` |
| `node`                             | `14.16.0` |

**Comment by bradzacher** [comment]
It looks like this is working as expected.

TypeScript is correctly inferring that `rule` is non-nullish because if your second try/catch throws, you never assign to the variable - meaning it retains its previous non-nullish value.

isolated repro without any imports
```ts
type RuleConfigValue = string | number | boolean | any[] | null
declare class MLRule<T extends RuleConfigValue, O = null> {}

declare function tryRequirePkg<T>(pkg: string, throwError: true): T;

export function moduleAutoLoader<T extends RuleConfigValue, O = unknown>(
) {
  const rules: Array<MLRule<T, O>> = []
  const errors: unknown[] = []

  let rule: MLRule<T, O> | null = null

  try {
    rule = tryRequirePkg<MLRule<T, O>>(`@markuplint/rule-${''}`, true)
  } catch (e) {
    errors.push(e)
  }

  if (!rule) {
    return
  }

  try {
    // if this throws - then `rule` is never assigned to, so it retains its non-null value
    rule = tryRequirePkg<MLRule<T, O>>(`markuplint-rule-${''}`, true)
  } catch (e) {
    errors.push(e)
  }

  if (!rule) { // reported here
    return
  }

  rules.push(rule)

  return {
    rules,
    errors,
  }
}
```

**Comment by JounQin** [comment]
@bradzacher Hah, right, the logic is incorrect actually. Thanks for pointing out that!
